src/Adapter/MSTest.TestAdapter/Helpers/ReflectHelper.cs

```diff
@@ -627,25 +462,64 @@ internal virtual IEnumerable<Trait> GetTestPropertiesAsTraits(MemberInfo testPro
                     description = string.Format(CultureInfo.CurrentCulture, Resource.ExceptionOccuredWhileGettingTheExceptionDescription, ex.GetType().FullName, ex2.GetType().FullName);                               // ex.GetType().FullName +
                 }
 
-                PlatformServiceProvider.Instance.AdapterTraceLogger.LogWarning(Resource.FailedToGetCustomAttribute, memberInfo.GetType().FullName!, description);
+                PlatformServiceProvider.Instance.AdapterTraceLogger.LogWarning(Resource.FailedToGetCustomAttribute, attributeProvider.GetType().FullName!, description);
 
-                // Since we cannot check by attribute names, do it in reflection way.
-                // Note 1: this will not work for different version of assembly but it is better than nothing.
-                // Note 2: we cannot cache this because we don't know if there are other attributes defined.
-                return null;
+                return Array.Empty<Attribute>();
             }
 
-            DebugEx.Assert(customAttributesArray != null, "attributes should not be null.");
+            DebugEx.Assert(attributes != null, "attributes should not be null.");
 
-            foreach (object customAttribute in customAttributesArray)
+            cache.Add(attributeProvider, attributes);
+
+            return attributes;
+        }
+    }
+
+    internal IEnumerable<TAttribute>? GetNonDerivedAttributes<TAttribute>(MethodInfo methodInfo, bool inherit)
+        where TAttribute : Attribute
+    {
+        Attribute[] cachedAttributes = GetCustomAttributesCached(methodInfo, inherit);
+
+        foreach (Attribute cachedAttribute in cachedAttributes)
+        {
+            if (AttributeComparer.IsNonDerived<TAttribute>(cachedAttribute))
             {
-                Type attributeType = customAttribute.GetType();
-                attributes[attributeType.AssemblyQualifiedName!] = customAttribute;
+                yield return (TAttribute)cachedAttribute;
             }
+        }
+    }
 
-            _attributeCache.Add(memberInfo, attributes);
+    /// <summary>
+    /// Reflection helper that is accessing Reflection directly, and won't cache the results.
+    /// </summary>
+    internal class NotCachedAttributeHelper : INotCachedAttributeHelper
+    {
+        /// <summary>
+        /// Get custom attributes on a member without cache.
+        /// </summary>
+        /// <param name="attributeProvider">Member for which attributes needs to be retrieved.</param>
+        /// <param name="inherit">If inherited type of attribute.</param>
+        /// <returns>All attributes of give type on member.</returns>
+        [return: NotNullIfNotNull(nameof(attributeProvider))]
+        public object[]? GetCustomAttributesNotCached(ICustomAttributeProvider attributeProvider, bool inherit)
```

Comment:
Maybe mark `attributeProvider` as nullable if we expect nulls?

Reply:
I am not even sure what I was doing here. removed, looks like it cannot be null,